### PR TITLE
Workbox

### LINKED ISSUE #250: Switch from sw-precache to workbox
What about use [`workbox`](https://github.com/GoogleChrome/workbox) instead of [`sw-precache`](https://github.com/GoogleChrome/sw-precache) in `polymer-build` to generate service worker code?

### PR DESCRIPTION
**Pull request by prateekbh**
Fixes #250

Start using `[workbox](https://github.com/GoogleChrome/workbox)` instead of `sw-precache` for service worker generation.

 - [ ] CHANGELOG.md has been updated

I am not sure if I have to add a new release or add it somewhere or how. please help around the CHANGELOG

### CODE REVIEW

File: src/service-worker.ts
Comment (prateekbh) [inline src/service-worker.ts]: this is something i am missing on right now, and would probably add as another manifestTransform

File: src/service-worker.ts
Comment (jeffposnick) [inline src/service-worker.ts]: The `workbox-build` node module exports a `getModuleUrl()` function, which will return the current CDN URL. I'd switch to that instead of hardcoding.

https://github.com/GoogleChrome/workbox/blob/ed2a847baf34c17c0f019d11cdfb1464fe3a9e19/packages/workbox-build/src/index.js#L86

(This is another method that is currently only used internally in the Webpack plugin, but it is supported.)

File: src/service-worker.ts
Comment (jeffposnick) [inline src/service-worker.ts]: I think you can get away with just setting

```js
workboxConfig.modifyUrlPrefix = {
  '': options.basePath
}
```

(Translation: replace the empty string as a prefix with the `options.basePath` value.)

File: src/service-worker.ts
Comment (jeffposnick) [inline src/service-worker.ts]: I think it's worth checking for `warnings` (which is an array of strings) in the `generateSWString()` response, and logging anything that's returned.

**Review by jeffposnick — CHANGES_REQUESTED** [review]
Thanks for working on this, @prateekbh!

I'll let the maintainers weigh in, but here are my comments from the Workbox perspective.

File: src/service-worker.ts
Comment (prateekbh) [reply src/service-worker.ts]: So I saw a concept of `bundled/unbundled` asset, which are passed to this method asn I just use them as exact globs to match and construct the precacheManifest

File: src/service-worker.ts
Comment (prateekbh) [reply src/service-worker.ts]: Done

File: src/service-worker.ts
Comment (prateekbh) [reply src/service-worker.ts]: TIL! Done

File: src/service-worker.ts
Comment (prateekbh) [reply src/service-worker.ts]: Done

File: src/service-worker.ts
Comment (prateekbh) [reply src/service-worker.ts]: Done

**Comment by justinfagnani** [comment]
Thanks @prateekbh, this is really helpful! I'll review this ASAP.

File: src/service-worker.ts
Comment (jeffposnick) [inline src/service-worker.ts]: Thanks for moving this into the `Object.assign()` logic.

In the back of my head, I'm wondering whether something special needs to be done for `importScripts`, apart from just overwriting the default `importScripts` array (which contains the Workbox CDN URL).

For instance, if `options.workboxConfig.importScripts === ['my-push-hander.js']`, I think the developer's intention is that they still want Workbox from the CDN, but they also want to load in their custom push handler script. With the code as it is right now, they'd just end up with the push handler and no Workbox CDN.

If we assume that the developer will *always* load Workbox from the CDN URL, then instead of overwriting the default `importScripts` in this `Object.assign()`, you could instead do

```js
workbox.config.importScripts =[getModuleUrl('workbox-sw')];
if ('importScripts' in options. workboxConfig) {
  workbox.config.importScripts = workbox.config.importScripts.concat(
    options.workboxConfig.importScripts);
}
```

Or something like that. It might not perfectly capture all use cases, but I think it gives some more flexibility—and if we get folks who ask for non-CDN builds, that would involve re-evaluating this anyway.

**Review by jeffposnick — CHANGES_REQUESTED** [review]

File: src/service-worker.ts
Comment (prateekbh) [reply src/service-worker.ts]: ah yes, makes perfect sense

Comment (prateekbh) [reply src/service-worker.ts]: Done!

**Comment by justinfagnani** [comment]
@prateekbh thanks again for all the work you've put in here. We've been quite busy working on Polymer 3.0 support in tools, which is the only reason why I haven't ben able to review this yet.

Since this PR was opened, we've moved to a mono at Polymer/tools. We'll need to port this PR over there.

@rictic, @graynorton and @kevinpschaaf putting this on your radar to look at if 3.0 issues are ironed out in the next week or so. This is a breaking change because the user-provided configs are different from sw-precache. We need to decide if we can support both workbox and sw-precache, or roll this into a set of CLI 2.0 breaking changes.

**Comment by prateekbh** [comment]
@rictic, @graynorton and @kevinpschaaf monorepo is kinda huge can you point me the location where the sw generation location exists.

**Comment by merlinnot** [comment]
https://github.com/Polymer/tools/blob/master/packages/build/src/service-worker.ts

**Pull request closed without merging**
